core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -609,6 +609,12 @@ class KafkaApis(val requestChannel: RequestChannel,
 
       mergedResponseStatus.forKeyValue { (topicPartition, status) =>
         if (status.error != Errors.NONE) {
+          if (status.error == Errors.PRODUCER_FENCED) {
```

Comment:
I feel this is not the right place to fix on broker: if we think produce request should never see a fenced producer exception, we should just modify `ProducerStateManager#checkProducerEpoch` to throw `InvalidProducerEpochException` instead.